Feat/stackerdb discovery

jcnelson opened the pull request:
This PR implements neighbor discovery for Stacker DB-aware Stacks nodes.  Stacks nodes report to each other via `StackerDBHandshakeAccept` which DBs they replicate (stored to the `LocalPeer` table in `PeerDB`; ultimately will be set by the config file).  Stacks nodes remember this information in the `PeerDB` as well, so that they can report which nodes replicate which DBs.

This PR also does some necessary refactoring work to the `NeighborWalk` state machine to separate out the low-level code for sending and receiving messages via the `PeerNetwork` from sending and reacting to messages.  This will be used in a subsequent PR to simplify the implementation of the Stacker DB state replication logic.

kantai commented on src/net/neighbors.rs:
```diff
@@ -1572,8 +1936,9 @@ impl NeighborWalk {
-        for (nkey, mut rh) in self.unresolved_neighbor_pings.drain() {
+        let mut unresolved_neighbor_pings =
+            std::mem::replace(&mut self.unresolved_neighbor_pings, HashMap::new());
+        for (nkey, mut rh) in unresolved_neighbor_pings.drain() {
```
What does `rh` stand for?

jcnelson replied:
Reply handle (`ReplyHandleP2P` specifically).

kantai reviewed (changes requested):
Thanks for these changes and the clearer refactoring in the network code. I had a few comments throughout, but I think the stackerdb changes need more testing coverage -- does the included testing handle updating stackerdb entries, deletion of a peer, inserting a peer on an existing slot, etc? I think those behaviors need unit test coverage.

stjepangolemac commented on src/net/db.rs:
```diff
@@ -363,6 +389,28 @@ const PEERDB_INITIAL_SCHEMA: &'static [&'static str] = &[
+    r#"
+    CREATE TABLE stackerdb_peers(
+        smart_contract_id TEXT NOT NULL,
+        peer_slot INTEGER NOT NULL,
```
I might be completely misunderstanding the question but here are my thoughts.

- slot seems to be calculated like this `hash(addr + ip)`
- according to [this](https://kevingal.com/apps/collision.html) the change for collision if there are 1000 peers is ~2.7089e-14
- is it possible for a different peer to get the same `(addr, port)` pair over time?

jcnelson replied:
Some background.

The Stacks node maintains a table called `frontier` which stores a record of every other peer it has ever contacted.  Maintaining this "phonebook" helps the peer recover from periods of intermittent network unavailability, including node crashes -- once the peer's network becomes available, it already has a persistent set of neighbors it can sample in order to re-establish itself as a node in the K-regular random peer graph.

Peers are discovered through the neighbor-walk state machine.  The neighbor-walk state machine itself is somewhat complex so I'll omit the details here, but the end result of the state-machine's execution is that it writes new rows to the `frontier` for newly-discovered neighbors.

Because peers can return any neighbor data they want, a malicious peer could take the opportunity to flood a victim's `frontier` table with garbage, leading to a DoS by resource exhaustion.  It can also report peer addresses that conflict with existing peers in the victim's `frontier` table.  To counter resource exhaustion, the the `frontier` table's size is necessarily size-bound -- in this case, up to 2**24 records may be inserted.

When inserting a peer into the frontier, each peer is hashed to eight different slots (rows).  If _at least one of them_ corresponds to an empty row in `frontier`, then the peer is inserted.  However, if _all eight_ slots are occupied, then the insert fails and the caller must first evict at least one colliding peer.

The neighbor-walk state machine is the only subsystem that ever inserts peers into the `frontier` table, and it handles the eviction case.  Specifically, it pings the peer to be evicted, and if that peer _does not respond_, then the newly-discovered peer is inserted.  Otherwise, the newly-discovered peer is dropped.  This approach is based on recommendations from [this USENIX paper](https://www.usenix.org/system/files/conference/usenixsecurity15/sec15-paper-heilman.pdf) on eclipse attacks on Bitcoin's p2p network.

---

The eviction strategy guarantees that `frontier.slot` values are unique.  A peer's row is not added if its slot ID is already present in the DB.  A peer's row can only be updated specifically in the case of peer eviction (in which case, the slot ID is preserved).

`peer_slot` is not a foreign key because `slot` cannot be a primary key.  Different peers can hash to the same slot IDs, but we would want to store both peers.

kantai replied:
Okay, but `slot` _is_ the primary key of the `frontier` table unless I'm really misreading something.

jcnelson replied:
`peer_slot` is now treated as a `FOREIGN KEY` to `frontier`

jcnelson commented on src/net/db.rs:
```diff
@@ -575,39 +675,60 @@ impl PeerDB {
+
+            tx.commit()?;
         }
-        if readwrite {
-            db.add_indexes()?;
-        }
```
Yes

jcnelson commented on src/net/db.rs:
```diff
@@ -1073,6 +1263,149 @@ impl PeerDB {
+        addrbytes: &PeerAddress,
+        port: u16,
+    ) -> Result<Vec<u32>, db_error> {
+        let slots = PeerDB::peer_slots(conn, network_id, addrbytes, port)?;
+        let mut ret = vec![];
+        for slot in &slots {
+            let used_slot = PeerDB::has_peer_at(conn, network_id, *slot)?;
```
See https://github.com/stacks-network/stacks-blockchain/pull/3552#discussion_r1159257436.  A peer has multiple slots in the DB.  This code is determining whether or not a given slot has a corresponding non-empty row in the DB.

kantai replied:
I am now kind of more confused by this method.

A peer has multiple slots where it _could_ land -- so `PeerDB::peer_slots` returns 8 (or whatever) results. Then `has_peer_at` in like 1297 just checks if there's _any_ peer at the returned slot.

But in the case of a peer B getting stored in one of peer A's alternates, wouldn't this method return a slot corresponding to a different peer? Why is this method returning a vec of slots at all?

And looking at how this is being used, it seems like this would be undesirable (i.e., Peer A being dropped doesn't mean that Peer B's "stackerdb information" should be deleted, right?)

kantai replied:
So, why couldn't this method be just the following?

`SELECT slot FROM frontier WHERE network_id = ? AND port = ? AND addrbytes = ?` 

The current method seems to be explicitly trying to include slots that the input peer _isn't_ occupying.

jcnelson replied:
Have a look at `try_insert_peer()`.  If _any_ slot generated by `PeerDB::peer_slots()` is _empty_, then the peer can be inserted and that particular slot marked as allocated.  If however there are no slots free, then the caller decides whether or not to evict one of the eight mapped peers from the frontier and replace it with the new peer.

You are indeed correct that this code is buggy.  When I wrote the Stacker DB code, I had forgotten to identify the _single_ slot that represents the peer to update.  Instead, the code erroneously updates all eight slots, even if they don't correspond to the same peer. Whoops!

kantai replied:
> Have a look at try_insert_peer(). If any slot generated by PeerDB::peer_slots() is empty, then the peer can be inserted and that particular slot marked as allocated. If however there are no slots free, then the caller decides whether or not to evict one of the eight mapped peers from the frontier and replace it with the new peer.

Yeah -- I understand that. My point was that the method `find_peer_slots()` appeared to be buggy, but perhaps also over-engineered. Why can't the `slot` field of `frontier` just be selected, and then `find_peer_slots()` returns a single slot? I understand a peer could be inserted in one of multiple slots, but it is never _actually resident_ in more than one, and from my understanding the intent of this method is to get the slot that a peer is actually resident in. Can you add a test coverage for the would-have-been buggy case here? (i.e., a test that would fail with the current implementation, but passes once this method is fixed?)

jcnelson replied:
Yes, you're correct.  The code now returns a single slot and has been renamed to `find_peer_slot()`.

jcnelson commented on src/net/db.rs:
```diff
@@ -1073,6 +1263,149 @@ impl PeerDB {
+            }
+        }
+
+        let sql = "DELETE FROM stackerdb_peers WHERE smart_contract_id = ?1";
+        for cid in to_delete.into_iter() {
+            test_debug!("Delete Stacker DB for {:?}: {}", &neighbor.addr, &cid);
+            let args: &[&dyn ToSql] = &[&cid];
+            tx.execute(sql, args).map_err(db_error::SqliteError)?;
+        }
```
Yes, every peer.  A peer can be represented up to `NUM_SLOT` times in the peer DB.

jcnelson commented on src/net/neighbors.rs:
```diff
@@ -241,6 +253,237 @@ impl NeighborWalkResult {
     }
 }
 
+/// A trait for representing a set of connected neighbors, for the purposes of executing some P2P
+/// algorithm.
+pub trait NeighborSet {
```
Yes, there are in the StackerDB PRs that build atop this one.  I put this trait here in order to _minimize_ the size of the other StackerDB PRs.

jcnelson commented on src/net/neighbors.rs:
```diff
@@ -241,6 +253,237 @@ impl NeighborWalkResult {
+        network: &mut PeerNetwork,
+        nk: &NeighborKey,
+    ) -> Result<ReplyHandleP2P, net_error> {
+        // send handshake.
```
It tries to move some bytes into the socket, and if the whole message has been moved into the socket, `try_send_recv()` succeeds.

jcnelson commented on src/net/neighbors.rs:
```diff
@@ -493,6 +736,94 @@ impl NeighborWalk {
+            }
+
+            let res = if neighbor_from_handshake.addr != self.cur_neighbor.addr {
+                // somehow, got a handshake from someone that _isn't_ cur_neighbor
+                debug!("{:?}: got unsolicited (or bootstrapping) HandshakeAccept from outbound {:?} (expected {:?})", 
+                           &self.local_peer,
+                           &neighbor_from_handshake.addr,
+                           &self.cur_neighbor.addr);
+
+                Err(net_error::PeerNotConnected)
```
No, this is a protocol violation (someone is spoofing `cur_neighbor`)

jcnelson commented on src/net/neighbors.rs:
```diff
@@ -493,6 +736,94 @@ impl NeighborWalk {
+            // connected to a routable neighbor, so update its entry in the DB.
+            let mut tx = network.peerdb.tx_begin()?;
+            let mut neighbor_from_handshake = Neighbor::from_handshake(
+                &mut tx,
```
`&mut` can be dropped.  `Neighbor::from_handshake()` takes a `&DBConn`; this `&mut tx` is getting deref'ed automatically.

jcnelson replied:
Done

kantai commented on src/net/db.rs:
```diff
@@ -456,21 +511,67 @@ impl PeerDB {
+    fn apply_schema_migrations<'a>(tx: &Transaction<'a>) -> Result<String, db_error> {
+        debug!("Apply any schema migrations");
+        let expected_version = PEERDB_VERSION.to_string();
+        let mut ret = None;
+        loop {
+            match PeerDB::get_schema_version(tx) {
+                Ok(version) => {
+                    if ret.is_none() {
+                        ret = Some(version.clone());
+                    }
+                    if version == "1" {
+                        PeerDB::apply_schema_2(tx)?;
+                    } else if version == expected_version {
+                        return Ok(ret.expect("unreachable"));
+                    } else {
+                        panic!("The schema version of the peer DB is invalid.")
+                    }
+                }
+                Err(e) => panic!("Error obtaining the version of the peer DB: {:?}", e),
+            }
+        }
```
I'd recommend eliminating nesting, the necessity of a mutable option for a non-optional return, and the use of `loop`:

```suggestion
        let begin_ver = PeerDB::get_schema_version(tx).expect("Error obtaining the version of the peer DB");
        let mut version = begin_ver.clone();
        while version != expected_version {
             version = PeerDB::get_schema_version(tx).expect("Error obtaining the version of the peer DB");
             if version == "1" {
                 PeerDB::apply_schema_2(tx)?;
             } else {
                 panic!("The schema version of the Peer DB is invalid. Found = {}", version);
             }
        }
       begin_ver
```

jcnelson replied:
Addressed

kantai commented on src/net/db.rs:
```diff
@@ -456,21 +511,67 @@ impl PeerDB {
+    }
+
+    fn apply_schema_migrations<'a>(tx: &Transaction<'a>) -> Result<String, db_error> {
+        debug!("Apply any schema migrations");
```
This debug line seems spurious. This will echo on every open of the PeerDB, correct? There's already other debug lines that echo whenever the PeerDB opens.

jcnelson replied:
Fixed

kantai commented on src/net/db.rs:
```diff
@@ -872,12 +1016,36 @@ impl PeerDB {
+        smart_contracts: &[ContractId],
+    ) -> Result<(), db_error> {
+        for cid in smart_contracts {
+            debug!("Add Stacker DB to slot {}: {}", slot, cid);
```
Can you consolidate this into a single debug line for the function call, rather than 1 per stacker db.

jcnelson replied:
This is now a `test_debug!`

jcnelson replied:
Fixed

kantai commented on src/net/db.rs:
```diff
@@ -1073,6 +1263,149 @@ impl PeerDB {
+        Ok(ret)
+    }
+
+    /// Get stacker DBs for a slot
```
Bump on this -- I think the usage of "stacker db" in the code currently is very confusing. It's being used ambiguously to refer both the data being tracked and the contract identifier of the contract governing that data. There are many places where a comment says something like "get stacker dbs for a slot" but it means something more like "get the ContractIDs whose associated databases are being tracked by the peers store in the slots `used_slots`" (because a "slot" is actually itself a reference). The methods and fields in this file need better rustdocs.

kantai commented on src/net/db.rs:
```diff
@@ -1532,6 +1933,103 @@ mod test {
     }
 
+    #[test]
+    fn test_insert_or_replace_stacker_dbs() {
```
For each of these new tests, can you add a comment describing what the test is intending to cover? What is the scenario being tested? Just like a sentence or two about what the test is supposed to be orchestrating would be helpful when figuring out what situation are being covered.

kantai reviewed (commented):
There were a number of unaddressed comments from the spring, and I had a few more comments as well.

jcnelson commented:
Sorry, I accidentally hit the re-request review button by accident. The
only thing that changed here was I merged the stacker db messages branch
into it.

On Mon, Jul 24, 2023, 5:37 PM Aaron Blankstein ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> There were a number of unaddressed comments from the spring, and I had a
> few more comments as well.
> ------------------------------
>
> In src/net/db.rs
> <https://github.com/stacks-network/stacks-blockchain/pull/3552#discussion_r1272755439>
> :
>
> > +        let mut ret = None;
> +        loop {
> +            match PeerDB::get_schema_version(tx) {
> +                Ok(version) => {
> +                    if ret.is_none() {
> +                        ret = Some(version.clone());
> +                    }
> +                    if version == "1" {
> +                        PeerDB::apply_schema_2(tx)?;
> +                    } else if version == expected_version {
> +                        return Ok(ret.expect("unreachable"));
> +                    } else {
> +                        panic!("The schema version of the peer DB is invalid.")
> +                    }
> +                }
> +                Err(e) => panic!("Error obtaining the version of the peer DB: {:?}", e),
> +            }
> +        }
>
> I'd recommend eliminating nesting, the necessity of a mutable option for a
> non-optional return, and the use of loop:
> ⬇️ Suggested change
>
> -        let mut ret = None;
> -        loop {
> -            match PeerDB::get_schema_version(tx) {
> -                Ok(version) => {
> -                    if ret.is_none() {
> -                        ret = Some(version.clone());
> -                    }
> -                    if version == "1" {
> -                        PeerDB::apply_schema_2(tx)?;
> -                    } else if version == expected_version {
> -                        return Ok(ret.expect("unreachable"));
> -                    } else {
> -                        panic!("The schema version of the peer DB is invalid.")
> -                    }
> -                }
> -                Err(e) => panic!("Error obtaining the version of the peer DB: {:?}", e),
> -            }
> -        }
> +        let begin_ver = PeerDB::get_schema_version(tx).expect("Error obtaining the version of the peer DB");
> +        let mut version = begin_ver.clone();
> +        while version != expected_version {
> +             version = PeerDB::get_schema_version(tx).expect("Error obtaining the version of the peer DB");
> +             if version == "1" {
> +                 PeerDB::apply_schema_2(tx)?;
> +             } else {
> +                 panic!("The schema version of the Peer DB is invalid. Found = {}", version);
> +             }
> +        }
> +       begin_ver
>
> ------------------------------
>
> In src/net/db.rs
> <https://github.com/stacks-network/stacks-blockchain/pull/3552#discussion_r1272758309>
> :
>
> > +            )
> +            .optional()?
> +            .unwrap_or("1".to_string());
> +        Ok(version)
> +    }
> +
> +    fn apply_schema_2<'a>(tx: &Transaction<'a>) -> Result<(), db_error> {
> +        debug!("Apply schema 2 to peer DB");
> +        for row_text in PEERDB_SCHEMA_2 {
> +            tx.execute_batch(row_text).map_err(db_error::SqliteError)?;
> +        }
> +        Ok(())
> +    }
> +
> +    fn apply_schema_migrations<'a>(tx: &Transaction<'a>) -> Result<String, db_error> {
> +        debug!("Apply any schema migrations");
>
> This debug line seems spurious. This will echo on every open of the
> PeerDB, correct? There's already other debug lines that echo whenever the
> PeerDB opens.
> ------------------------------
>
> In src/net/db.rs
> <https://github.com/stacks-network/stacks-blockchain/pull/3552#discussion_r1272760532>
> :
>
> > @@ -872,12 +1016,36 @@ impl PeerDB {
>          Ok(allow_rows)
>      }
>
> +    /// Insert or replace stacker DBs for a peer, given its slot
> +    pub fn insert_or_replace_stacker_dbs<'a>(
> +        tx: &mut Transaction<'a>,
> +        slot: u32,
> +        smart_contracts: &[ContractId],
> +    ) -> Result<(), db_error> {
> +        for cid in smart_contracts {
> +            debug!("Add Stacker DB to slot {}: {}", slot, cid);
>
> Can you consolidate this into a single debug line for the function call,
> rather than 1 per stacker db.
> ------------------------------
>
> In src/net/db.rs
> <https://github.com/stacks-network/stacks-blockchain/pull/3552#discussion_r1272783493>
> :
>
> > @@ -1532,6 +1933,103 @@ mod test {
>          }
>      }
>
> +    #[test]
> +    fn test_insert_or_replace_stacker_dbs() {
>
> For each of these new tests, can you add a comment describing what the
> test is intending to cover? What is the scenario being tested? Just like a
> sentence or two about what the test is supposed to be orchestrating would
> be helpful when figuring out what situation are being covered.
> ------------------------------
>
> In src/net/db.rs
> <https://github.com/stacks-network/stacks-blockchain/pull/3552#discussion_r1272778433>
> :
>
> > @@ -363,6 +389,28 @@ const PEERDB_INITIAL_SCHEMA: &'static [&'static str] = &[
>  const PEERDB_INDEXES: &'static [&'static str] =
>      &["CREATE INDEX IF NOT EXISTS peer_address_index ON frontier(network_id,addrbytes,port);"];
>
> +const PEERDB_SCHEMA_2: &'static [&'static str] = &[
> +    r#"
> +    CREATE TABLE stackerdb_peers(
> +        smart_contract_id TEXT NOT NULL,
> +        peer_slot INTEGER NOT NULL,
>
> Okay, but slot *is* the primary key of the frontier table unless I'm
> really misreading something.
> ------------------------------
>
> In src/net/db.rs
> <https://github.com/stacks-network/stacks-blockchain/pull/3552#discussion_r1272774463>
> :
>
> > +        let slots = PeerDB::peer_slots(conn, network_id, addrbytes, port)?;
> +        let mut ret = vec![];
> +        for slot in &slots {
> +            let used_slot = PeerDB::has_peer_at(conn, network_id, *slot)?;
>
> I am now kind of more confused by this method.
>
> A peer has multiple slots where it *could* land -- so PeerDB::peer_slots
> returns 8 (or whatever) results. Then has_peer_at in like 1297 just
> checks if there's *any* peer at the returned slot.
>
> But in the case of a peer B getting stored in one of peer A's alternates,
> wouldn't this method return a slot corresponding to a different peer? Why
> is this method returning a vec of slots at all?
>
> And looking at how this is being used, it seems like this would be
> undesirable (i.e., Peer A being dropped doesn't mean that Peer B's
> "stackerdb information" should be deleted, right?)
> ------------------------------
>
> In src/net/db.rs
> <https://github.com/stacks-network/stacks-blockchain/pull/3552#discussion_r1272780587>
> :
>
> > +        let slots = PeerDB::peer_slots(conn, network_id, addrbytes, port)?;
> +        let mut ret = vec![];
> +        for slot in &slots {
> +            let used_slot = PeerDB::has_peer_at(conn, network_id, *slot)?;
>
> So, why couldn't this method be just the following?
>
> SELECT slot FROM frontier WHERE network_id = ? AND port = ? AND addrbytes
> = ?
>
> The current method seems to be explicitly trying to include slots that the
> input peer *isn't* occupying.
> ------------------------------
>
> In src/net/db.rs
> <https://github.com/stacks-network/stacks-blockchain/pull/3552#discussion_r1272765086>
> :
>
> > +        network_id: u32,
> +        addrbytes: &PeerAddress,
> +        port: u16,
> +    ) -> Result<Vec<u32>, db_error> {
> +        let slots = PeerDB::peer_slots(conn, network_id, addrbytes, port)?;
> +        let mut ret = vec![];
> +        for slot in &slots {
> +            let used_slot = PeerDB::has_peer_at(conn, network_id, *slot)?;
> +            if used_slot {
> +                ret.push(*slot);
> +            }
> +        }
> +        Ok(ret)
> +    }
> +
> +    /// Get stacker DBs for a slot
>
> Bump on this -- I think the usage of "stacker db" in the code currently is
> very confusing. It's being used ambiguously to refer both the data being
> tracked and the contract identifier of the contract governing that data.
> There are many places where a comment says something like "get stacker dbs
> for a slot" but it means something more like "get the ContractIDs whose
> associated databases are being tracked by the peers store in the slots
> used_slots" (because a "slot" is actually itself a reference). The
> methods and fields in this file need better rustdocs.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/stacks-network/stacks-blockchain/pull/3552#pullrequestreview-1544275443>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADQJK5GIUZGBC4Z5RAGL73XR3TKNANCNFSM6AAAAAAURREFT4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

jcnelson commented on src/net/db.rs:
```diff
@@ -1073,6 +1263,149 @@ impl PeerDB {
+            if !new_dbs_table.contains(cid) {
+                to_delete.push(cid);
+            }
+        }
```
Done

jcnelson commented on src/net/neighbors.rs:
```diff
@@ -241,6 +253,237 @@ impl NeighborWalkResult {
+                        return Err(Err(e));
+                    }
+                }
+            }
```
That file no longer exists, and has been refactored into `src/net/neighbors/` (which has a lot less indentation).

jcnelson commented on src/net/neighbors/comms.rs:
N.B. This is part of refactoring `src/net/neighbors.rs` into multiple files.

jcnelson commented on src/net/neighbors/mod.rs:
N.B. This is part of refactoring `src/net/neighbors.rs` into multiple files.

jcnelson commented on src/net/neighbors/neighbor.rs:
N.B. This is part of refactoring `src/net/neighbors.rs` into multiple files

jcnelson commented on src/net/neighbors/walk.rs:
N.B. This is part of refactoring `src/net/neighbors.rs` into multiple files

jcnelson commented on src/net/tests/neighbors.rs:
N.B. This is part of refactoring `src/net/neighbors.rs` into multiple files

jcnelson commented:
@kantai An update here.  I've gotten the ball rolling on refactoring the `NeighborWalk` state machine to separate database I/O and transport I/O from the walk logic.  This has resulted in splitting `src/net/neighbors` into files `src/net/neighbors/*.rs` and `src/net/tests/neighbors.rs`.  I don't think the refactoring is "done" insofar as the abstractions are perfect, but this diff has gotten quite large and I'm reluctant to pursue it further here.  At the same time, I've reduced levels of indentation wherever I could. 
 IMHO the neighbor-walk logic is a lot cleaner and easier to grasp now (but it will be made better in the future!)
 
 I can continue to add unit tests to the now-isolated DB I/O helpers and state-machine steps, but this would be outside of this PR's intended scope.  I can do so, but it will take an additional day or two, when accounting for non-coding hours, and it will make the diff larger.  What do you think?
 
 My ultimate goal here is to create a `NetworkStateMachine` trait that implements a ton of helper methods for handling transport-level concerns, and refactor the `InvSync`, `BlockDownloader`, mempool sync, and microblock sync state machines to use it.  This trait would serve to isolate the business logic of these state machines from the `PeerNetwork` instance, allowing the networking nuts and bolts to be mocked by a future property tester or fuzzer.  Once I get to finish `feat/stackerdb-sync`, I'll have an opportunity to take a crack at creating this trait, since I can then apply it to both `NeighborWalk` and the StackerDB sync system (and be confident that the abstraction is sufficiently expressive to permit the aforementioned refactorings).
 
This not only serves the quality-of-life workstream, but also the better blocks workstream:  by having a state machine trait that isolates state-machine logic from the underlying networking system, it becomes tenable to create the new block inventory sync and block downloader state machines without having to become an expert in the networking stack first.  I'm hoping that this will allow these tasks to be taken off my plate.

jcnelson commented on src/net/neighbors.rs:
```diff
@@ -493,6 +736,94 @@ impl NeighborWalk {
+                &self.local_peer, &self.cur_neighbor.addr, &neighbor_from_handshake.addr
+            );
+
+            self.neighbor_from_handshake = neighbor_from_handshake.addr;
```
Done

kantai commented on src/net/db.rs:
```diff
@@ -1073,6 +1283,136 @@ impl PeerDB {
+        addrbytes: &PeerAddress,
+        port: u16,
+    ) -> Result<Option<u32>, db_error> {
+        // the peer can be at any of these slots, if it is present at all.
+        let slots = PeerDB::peer_slots(conn, network_id, addrbytes, port)?;
+        for slot in slots.into_iter() {
+            let used_slot = PeerDB::has_specific_peer_at(conn, network_id, addrbytes, port, slot)?;
+            if used_slot {
+                return Ok(Some(slot));
+            }
+        }
```
Why does this function need to iterate? Can't the occupied slot, if the peer is present, be queried directly?


```
"SELECT slot FROM frontier WHERE network_id = ?1 AND addrbytes = ?2 AND port = ?3";
```

jcnelson replied:
Yup, will fix

kantai commented on src/net/db.rs:
```diff
@@ -1073,6 +1283,136 @@ impl PeerDB {
+        let to_insert: Vec<_> = new_dbs_set.difference(&cur_dbs_set).collect();
+        let to_delete: Vec<_> = cur_dbs_set.difference(&new_dbs_set).collect();
+
+        let sql = "DELETE FROM stackerdb_peers WHERE smart_contract_id = ?1";
+        for cid in to_delete.into_iter() {
+            test_debug!("Delete Stacker DB for {:?}: {}", &neighbor.addr, &cid);
+            let args: &[&dyn ToSql] = &[&cid.to_string()];
+            tx.execute(sql, args).map_err(db_error::SqliteError)?;
+        }
```
This deletes those stacker db entries for _all peers_, is that the intended behavior of this function?

jcnelson replied:
Nope, will fix

kantai commented on src/net/db.rs:
```diff
@@ -1073,6 +1283,136 @@ impl PeerDB {
+        neighbor: &Neighbor,
+        dbs: &[ContractId],
+    ) -> Result<(), db_error> {
+        let cur_slot_opt = PeerDB::find_peer_slot(
+            tx,
+            neighbor.addr.network_id,
+            &neighbor.addr.addrbytes,
+            neighbor.addr.port,
+        )?;
```
If there's no slot, shouldn't this function do nothing?

```suggestion
        let neighbor_slot = match PeerDB::find_peer_slot(
            tx,
            neighbor.addr.network_id,
            &neighbor.addr.addrbytes,
            neighbor.addr.port,
        )? {
            Some(x) => x,
            None => return Ok(()),
        };
```

jcnelson replied:
Sure thing

kantai commented on src/net/db.rs:
```diff
@@ -1073,6 +1283,136 @@ impl PeerDB {
+    /// Calculates the delta between what's in the DB now, and what's in `dbs`, and deletes the
+    /// records absent from `dbs` and adds records not present in the DB.
+    /// Does nothing if the peer is not present.
+    pub fn update_peer_stacker_dbs<'a>(
+        tx: &Transaction<'a>,
```
Here, and almost anywhere else `&Transaction<'a>` is used as an argument, the lifetime parameter is unnecessary and should be removed.

jcnelson replied:
Will do

kantai commented on src/net/neighbors/db.rs:
```diff
@@ -0,0 +1,652 @@
+        &self,
+        network: &mut PeerNetwork,
+        replacements: &NeighborReplacements,
+        result: &mut NeighborWalkResult,
+    ) -> Result<(), net_error> {
+        Self::inner_replace_neighbors(network, replacements, result)
+    }
+
+    fn neighbor_from_handshake(
+        &self,
+        network: &PeerNetwork,
+        preamble: &Preamble,
+        data: &HandshakeAcceptData,
+    ) -> Result<Neighbor, net_error> {
+        Self::inner_neighbor_from_handshake(network, preamble, data)
+    }
+
+    fn save_neighbor_from_handshake(
+        &self,
+        network: &mut PeerNetwork,
+        preamble: &Preamble,
+        data: &HandshakeAcceptData,
+        db_data: Option<&StackerDBHandshakeData>,
+    ) -> Result<Neighbor, net_error> {
+        Self::inner_save_neighbor_from_handshake(network, preamble, data, db_data)
+    }
+
+    fn update_neighbor(
+        &self,
+        network: &mut PeerNetwork,
+        cur_neighbor: Neighbor,
+        new_data: Option<&HandshakeAcceptData>,
+        new_db_data: Option<&StackerDBHandshakeData>,
+    ) -> Result<Neighbor, net_error> {
+        Self::inner_update_neighbor(network, cur_neighbor, new_data, new_db_data)
+    }
+
+    fn get_asn_count(&self, network: &PeerNetwork, asn: u32) -> u64 {
+        Self::inner_get_asn_count(network, asn)
+    }
```
Why add all these one-line shuttle invocations of `inner_*`? It seems like those methods could all just be implemented right in the trait implementation.

jcnelson replied:
Sure, I can do this.  I just like having my implementation separate from the trait binding, but that's really not needed here.

kantai commented on src/net/neighbors/db.rs:
```diff
@@ -0,0 +1,652 @@
+            };
+
+            let replaced_opt = PeerDB::get_peer_at(&mut tx, network_id, *slot)?;
+            if let Some(replaced) = replaced_opt {
+                if PeerDB::is_address_denied(&mut tx, &replacement.addr.addrbytes)? {
+                    debug!(
+                        "{:?}: Will not replace {:?} with {:?} -- is denied",
+                        local_peer_str, &replaced.addr, &replacement.addr
+                    );
+                } else {
+                    debug!(
+                        "{:?}: Replace {:?} with {:?}",
+                        local_peer_str, &replaced.addr, &replacement.addr
+                    );
+
+                    PeerDB::insert_or_replace_peer(&mut tx, &replacement, *slot)?;
+                    result.add_replaced(replaced.addr.clone());
+                }
```
Does this _only_ insert `replacement` if the slot was occupied? What if the slot was vacant? Shouldn't the new neighbor still be inserted?

kantai commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2003 @@
+        );
+
+        // find an inbound connection
+        for _ in 0..network.get_walk_pingbacks().len() + 1 {
+            let event_id = match network.iter_peer_event_ids().skip(idx).next() {
+                Some(eid) => *eid,
+                None => {
+                    idx = 0;
+                    continue;
+                }
+            };
+            idx = (idx + 1) % network.get_num_p2p_convos();
```
This loop is confusing to me -- why are you iterating `get_walk_pingbacks().len()` times, but over the number of conversations? Are those items the same length?

kantai commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2003 @@
+            return Ok(false);
+        };
+
+        match message.payload {
+            StacksMessageType::HandshakeAccept(ref data) => {
+                debug!(
+                    "{:?}: received HandshakeAccept from {} {:?}: {:?}",
+                    network.get_local_peer(),
+                    if self.walk_outbound {
+                        "outbound"
+                    } else {
+                        "inbound"
+                    },
+                    &message.to_neighbor_key(&data.handshake.addrbytes, data.handshake.port),
+                    &data.handshake
+                );
+
+                self.handle_handshake_accept(network, &message.preamble, data, None)?;
+            }
+            StacksMessageType::StackerDBHandshakeAccept(ref data, ref db_data) => {
+                debug!(
+                    "{:?}: received StackerDBHandshakeAccept from {} {:?}: {:?}, {:?}",
+                    network.get_local_peer(),
+                    if self.walk_outbound {
+                        "outbound"
+                    } else {
+                        "inbound"
+                    },
+                    &message.to_neighbor_key(&data.handshake.addrbytes, data.handshake.port),
+                    &data.handshake,
+                    db_data
+                );
+
+                self.handle_handshake_accept(network, &message.preamble, data, Some(db_data))?;
+            }
```
Thes `handshake_accepts` seem like they could be deduped by doing something like:

```
let (handshake_data, stacker_db_data) = match message.payload {
    HandshakeAccept(ref data) => (data, None),
    StackerDBHandshakeAccept(ref data, ref db_data) => (data, Some(db_data)),
     ...
};
```

kantai commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2003 @@
+        for (naddr, message) in self.comms.collect_replies(network).into_iter() {
+            let nkey = naddr.to_neighbor_key(network);
+            match message.payload {
+                StacksMessageType::HandshakeAccept(ref data) => {
+                    debug!(
+                        "{:?}: Got HandshakeAccept from {:?}",
+                        network.get_local_peer(),
+                        &nkey
+                    );
+
+                    self.handle_neighbor_handshake_accept(
+                        network,
+                        naddr,
+                        &message.preamble,
+                        data,
+                        None,
+                    )?;
+                }
+                StacksMessageType::StackerDBHandshakeAccept(ref data, ref db_data) => {
+                    debug!(
+                        "{:?}: Got StackerDBHandshakeAccept from {:?}: {:?}",
+                        network.get_local_peer(),
+                        &nkey,
+                        db_data
+                    );
+
+                    self.handle_neighbor_handshake_accept(
+                        network,
+                        naddr,
+                        &message.preamble,
+                        data,
+                        Some(db_data),
+                    )?;
+                }
+                StacksMessageType::HandshakeReject => {
```
Same as elsewhere, the handshakeaccept handling should be deduped.

kantai commented on src/net/neighbors/db.rs:
```diff
@@ -0,0 +1,652 @@
+        if let Some(slot) = replaced_neighbor_slot_opt {
+            // if this peer isn't allowed or denied, then consider
+            // replacing.  Otherwise, keep the local configuration's preference.
+            if !neighbor_from_handshake.is_denied() && !neighbor_from_handshake.is_allowed() {
```
Why should an allowed peer not get added?

jcnelson replied:
We don't want to have the neighbor walk algorithm override local policies regarding whether or not a node is set as allowed by the operator, or set as denied by the operator.  In either case, the peer's record needs to stick around in the DB.

kantai commented on src/net/neighbors/db.rs:
```diff
@@ -0,0 +1,652 @@
+                PeerDB::get_peer_at(conn, nk.network_id, slot).map_err(net_error::DBError)?;
+
+            match peer_opt {
+                None => {
+                    continue;
+                }
```
Why `continue` here? If the node finds an empty slot, shouldn't it just insert there?

jcnelson replied:
Yup; I don't think this is actually reachable though.  But yes -- either way, we pick a slot at random

kantai reviewed (commented):
I have some comments on the latest changes. I still haven't reviewed neighbors/comms.rs, neighbors/mod.rs, or tests/neighbors.rs. I'll try to get through them today as well, but I wanted to submit my comments now because some of the changes since the review was requested ended up requiring I go through them again.

kantai commented on src/net/neighbors/mod.rs:
```diff
@@ -0,0 +1,430 @@
+        );
+
+        // always ensure we're connected to always-allowed outbound peers
+        let walk_res = if ibd {
+            // always connect to bootstrap peers if in IBD
+            NeighborWalk::instantiate_walk_to_always_allowed(
+                self.get_neighbor_walk_db(),
+                self.get_neighbor_comms(),
+                self,
+                ibd,
+            )
+        } else {
+            // if not in IBD, then we're not required to use the always-allowed neighbors
+            // all the time (since they may be offline, and we have all the blocks anyway).
+            // Alternate between picking random neighbors, and picking always-allowed
+            // neighbors.
+            if self.walk_attempts % (self.connection_opts.walk_inbound_ratio + 1) == 0 {
+                NeighborWalk::instantiate_walk(
+                    self.get_neighbor_walk_db(),
+                    self.get_neighbor_comms(),
+                    self,
+                )
+            } else {
+                NeighborWalk::instantiate_walk_to_always_allowed(
+                    self.get_neighbor_walk_db(),
+                    self.get_neighbor_comms(),
+                    self,
+                    ibd,
+                )
+            }
+        };
```
```suggestion
        // always connect to bootstrap peers if in IBD. if not, alternate between random peers
        // and always connected ones 
        let walk_to_always_connected_peer = ibd || (self.walk_attempts % (self.connection_opts.walk_inbound_ratio + 1) != 0);
        let walk_res = if walk_to_always_connected_peer {
            NeighborWalk::instantiate_walk_to_always_allowed(
                self.get_neighbor_walk_db(),
                self.get_neighbor_comms(),
                self,
                ibd,
            )
        } else {
            NeighborWalk::instantiate_walk(
                    self.get_neighbor_walk_db(),
                    self.get_neighbor_comms(),
                    self,
                )
        };
```

jcnelson replied:
sure

jcnelson replied:
I'm going to tweak this slightly.  What I want to happen is if the node isn't IBD, then it should alternate between outbound walks, inbound walks, and pingback walks.

kantai commented on src/net/neighbors/mod.rs:
```diff
@@ -0,0 +1,430 @@
+
+        let walk_result_opt = match walk.run(self) {
+            Ok(Some(ws)) => {
+                // walk ran to completion
+                self.walk_count += 1;
+                self.walk_deadline = self.connection_opts.walk_interval + get_epoch_time_secs();
+
+                debug!(
+                    "{:?}: walk has completed in {} steps ({} walks total)",
+                    &self.local_peer, walk.walk_step_count, self.walk_count
+                );
+
+                self.print_walk_diagnostics();
+                Some(ws)
```
This branch returns `Some(ws)` if the walk ran to completion, and it is the only case in the match that continues execution of `walk_peer_graph`, but later on this function, the node:

1. checks if the walk timed out (which it shouldn't have if the walk completed, but also, if the walk completed before this later check, why should it time out now after the fact)
2. checks if there should be an occasional walk restart

If the walk timed out or should be restarted, this method returns `(true, None)` indicating that there's no more work to do on the walk. But if the walk completed, didn't time out, and didn't get the ocassional restart, this method returns `(false, Some(walk_result))`, indicating that work _does_ need to be done, even though the walk is finished. I guess I'm mostly confused about that last case: why is the boolean in that case `false`? Is there more work that needs to be done on the walk?

jcnelson replied:
Ah, I think the `Ok(None)` case for `walk.run()` should _not_ return.  We want the timeout check to apply on each pass as long as there's not an error.

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+    }
+
+    /// Get a random starting neighbor for the walk.
+    /// Older but still fresh neighbors will be preferred.
```
I have since elaborated on this.

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+        let mut neighbor_from_handshake =
+            Neighbor::from_handshake(tx, peer_version, network_id, handshake)?;
+
+        if Neighbor::from_neighbor_address(tx, network_id, block_height, naddr)?.is_some() {
```
In this PR, I'll `Neighbor::from_handshake()` to return `(Neighbor, bool)` to indicate whether or not the returned `Neighbor` represents DB-backed state.  This would obviate the call to `Neighbor::from_neighbor_address()`.

kantai replied:
I don't think the return types need to change at all here. Simple renaming would make it much more clear what these methods are doing. I'd suggest renaming `from_handshake()` to `Neighbor::load_and_update()` -- the "handshake" is implicit anyways because it accepts handshake data as a parameter, and what the method is doing is "loading" and "updating" (the method itself invokes a "get" and then "update" method).

And then I'd suggest renaming `from_neighbor_address()` to `load_by_address` or similar -- it's not "constructing a neighbor from an address" (which is what `from_neighbor_address` would imply), it's loading it from a db.

jcnelson replied:
Sure, I can do that

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+            ret.push(neighbor);
+        }
+        ret
+    }
```
Will do

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+        if let Some(slot) = replaced_neighbor_slot_opt {
+            // if this peer isn't allowed or denied, then consider
+            // replacing.  Otherwise, keep the local configuration's preference.
+            if !neighbor_from_handshake.is_denied() && !neighbor_from_handshake.is_allowed() {
```
I fixed `is_allowed()` and `is_denied()` so that they're now mutually exclusive.  Now, this check only uses `.is_denied()`.

kantai replied:
I think this check just moved to `neighbors/db.rs:401`.

My confusion is why a new neighbor that is "allowed" should be treated the same as a new neighbor that is "denied". The if statement here makes it so that a replacement is only scheduled if the new neighbor is denied _or_ the new neighbor is allowed.

jcnelson replied:
Notions of "allowed" and "denied" in this context refer to local policy.  A node is "allowed" if the operator explicitly marked it as such in the PeerDB (e.g. through a config option), meaning that it should never be banned or blocked by the network.  Similarly, a node is "denied" if the operator explicitly marked it as such, or if it got banned and was not "allowed."

Under these terms, a node can be both "allowed" and "denied" through a bad local configuration, which really ought to never happen.  The point of this code was to avoid overriding the node operator's preferences -- if the node operator explicitly allowed or denied a node, we don't want to replace its record in the PeerDB (because we want those policies to stay that way).

kantai replied:
But isn't a neighbor only in this code path if it isn't already stored in the peer db? So is this a dead check (i.e., there's no way for a neighbor to return true for "allowed" and the code to get to this line, because it would have returned somewhere around the check `Neighbor::from_neighbor_address(&mut tx, preamble.network_id, block_height, naddr)?.is_some()`?

jcnelson replied:
No, because `is_allowed()` is a function of the `allowed` value in `Neighbor`, which is a UNIX timestamp.  A neighbor can be marked as allowed or denied until a certain time (in the latter case, we use the timestamp to permit bans to expire).  A neighbor can be loaded via `Neighbor::from_neighbor_address()` if its `allowed` value is in the past (meaning, `is_allowed()` is false).

kantai replied:
I'm sorry to keep circling around on this point, but I think I'm still pretty confused here.

Here is the start of the method we're talking about:

```rust
        let mut neighbor_from_handshake = Neighbor::from_handshake(
            &mut tx,
            preamble.peer_version,
            preamble.network_id,
            handshake,
        )?;
```

So at the start of the method, `neighbor_from_handshake` is initialized either (1) entirely from the handshake data or (2) from the db and then updated with the handshake data. My understanding, based on what you're saying, is that the _only_ way `is_allowed()` could be true is if the case was (2).

Immediately after that, the method then does this check:

```rust
        if Neighbor::from_neighbor_address(&mut tx, preamble.network_id, block_height, naddr)?
            .is_some()
        {
            test_debug!("already know about {:?}", naddr);
            neighbor_from_handshake
                .save_update(&mut tx, db_data.map(|x| x.smart_contracts.as_slice()))?;
            tx.commit()?;

            // seen this neighbor before
            return Ok((false, neighbor_from_handshake));
        }
```

`Neighbor::from_neighbor_address` returns `Some` iff the neighbor identified by (`preamble.network_id`, `naddr` ) is in the db. I believe that `naddr` and the (`addrbytes`, `port`) in `handshake` are the same (i.e., `naddr` is a reference to the neighbor that returned the `handshake` data). In which case, `Neighbor::from_neighbor_address` returns `Some` if `neighbor_from_handshake` was loaded from the db (i.e., case (2) above). If it returns `Some`, then the method _will return_ at this point.

So, if `neighbor_from_handshake.is_allowed()` is true, then `neighbor_from_handshake` was loaded from the db. If it was loaded from the db, then this if statement executes. If this if statement executes, the method returns before ever getting to the save or replace logic.

jcnelson replied:
> (2) from the db and then updated with the handshake data. My understanding, based on what you're saying, is that the only way is_allowed() could be true is if the case was (2).

This is not true.  `is_allowed()` can return false but the neighbor will be loaded.  This is because `neighbor_from_handshake` uses `PeerDB::get_peer()`, which doesn't look at allowed or denied policy.  `PeerDB::get_peer()` (and thus this method) will load a neighbor that is banned.

The `PeerNetwork` will not connect to a denied peer, however.

kantai replied:
But it doesn't matter if `is_allowed()` _could_ return false, the check in question is guarding against `is_allowed` being true:

```
            if !neighbor_from_handshake.is_denied() && !neighbor_from_handshake.is_allowed() {
                replacements.add_neighbor(
                    NeighborAddress::from_neighbor(&neighbor_from_handshake),
                    neighbor_from_handshake.clone(),
                    slot,
                );
            } else {
                unreachable!("My point is that this hypothetical case being guarded against seems unreachable.");
            }
```

kantai replied:
If either `is_denied()` or `is_allowed()` is true, then `neighbor_from_handshake` was loaded from the database (I'm not saying that the inverse is true: I understand that these methods can return false for data loaded from the database).

If `neighbor_from_handshake` was loaded from the database, then the method returns before ever performing that `!is_denied() && !is_allowed()` check, because it returns in the branch corresponding to the `test_debug!("already know about {:?}", naddr);`

jcnelson replied:
Oh, I see now.  Yes, it appears that this is a dead check, because as you say, if either `is_denied()` or `is_allowed()` is `true`, then the `allowed` and/or `denied` values must have been loaded from the DB since they'd be non-zero.  I'll remove the check.

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+        Ok(None)
+    }
+
+    /// Add a neighbor or schedule it to be pinged since it's up for replacement.
```
Yes, that's correct.  If we have a new neighbor to add, but all of its slots are taken, then we can take one of two actions:

* Replace the old neighbor with the new neighbor
* Keep the old neighbor and drop the new neighbor

The code will prefer keeping old neighbors if they're still alive, since it prevents an evil neighbor eclipsing it by flooding them with garbage neighbor state.  To do this, the state machine will ping the old neighbor to see if it responds.  If it does, the new neighbor is dropped.  If it doesn't, the old neighbor is replaced.  This means that in order for an evil node to eclipse this node, it would have to set up lots and lots of real nodes for this node to handshake with as well as somehow sever all links to the old nodes.

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+            return Err(net_error::NotFoundError);
+        };
+
+        debug!(
+            "Will (re-)connect to always-allowed peer {:?}",
+            &allowed_peer.addr
+        );
```
Will do

kantai commented on src/net/neighbors/db.rs:
```diff
@@ -0,0 +1,652 @@
+    }
+
+    pub fn add_neighbor(&mut self, naddr: NeighborAddress, neighbor: Neighbor, slot: u32) {
+        self.replacements.insert(naddr.clone(), neighbor);
+        self.replaced_neighbors.insert(naddr, slot);
```
What if `slot` has already been inserted here? That is, the same neighbor is slated to be replaced by two new neighbors. I don't think that leads to any incoherence in the peer state, because the second new neighbor just replaces the first new neighbor during `replace_neighbors()`, but is that the intended behavior?

jcnelson replied:
It's not intentional, but I don't think it warrants a fix.  Slot collisions are highly unlikely but not astronomically so -- per the birthday paradox, you'd need about 2^16 nodes inserted before the probability of a slot collision exceeds 50%.  Even if there's a collision, there's no loss of coherence.

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+                    continue;
+                }
+            };
+            idx = (idx + 1) % network.get_num_p2p_convos();
```
You're right.  Will fix.

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+        }
+
+        // if cur_neighbor is _us_, then grab a different neighbor and try again
+        if Hash160::from_node_public_key(&self.cur_neighbor.public_key)
+            == Hash160::from_node_public_key(&Secp256k1PublicKey::from_private(
+                &network.get_local_peer().private_key,
+            ))
```
The function `Hash160::from_node_public_key()` evaluates to the hash160 of the compressed public key, even if the argument is not compressed.  It's defensive programming.

kantai replied:
That's a pretty surprising usage of these methods -- that should either have a comment like:

```rust
// compare the Hash160's because `Hash160::from_node_public_key` forces comparison of the compressed public // keys
```

So that a future maintainer knows why this is being done, or it should be replaced with an explicit force-to-compressed-key method.

jcnelson replied:
I'll add a comment.

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+                self.add_broken(network, &self.cur_neighbor.addr.clone());
+                Err(net_error::InvalidMessage)
+            }
+        }
```
Sure, will do

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+            }
+        };
+
+        let mut neighbor_addrs = match message.payload {
```
Sure, will do.

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+            ));
+        }
+
+        // prune the list to a reasonable size in case cur_neighbor gave us too many for our
+        // configuration
+        if neighbor_addrs.len() as u64 > network.get_connection_opts().max_neighbors_of_neighbor {
```
I'm not sure how whether or not the peer is inbound is relevant here?  This is to prevent a denial of service attack.  Suppose the remote node gave me a 16MB list of neighbors, each of which took a long time to respond.

kantai replied:
Immediately before this pruning, the node checks if it is doing an inbound walk, and adds the cur_neighbor to the to_resolve list:

```
        if !self.walk_outbound {
            test_debug!("{:?}: will try to handshake with inbound neighbor {:?}'s advertized address {:?} as well", network.get_local_peer(), &self.cur_neighbor.addr, &self.neighbor_from_handshake);
            let cur_neighbor_pubkey_hash =
                Hash160::from_node_public_key(&self.cur_neighbor.public_key);
            neighbor_addrs.push(NeighborAddress::from_neighbor_key(
                self.neighbor_from_handshake.clone(),
                cur_neighbor_pubkey_hash,
            ));
        }
```

Pruning _after_ that means that a neighbor could return a maximally valid list, then this node would add `cur_neighbor` to the "to resolve" list, and prune the `cur_neighbor`.

jcnelson replied:
Ah, yes. You're right; pruning before would be better here.  Will change.

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+        let my_pubkey_hash = Hash160::from_node_public_key(&Secp256k1PublicKey::from_private(
+            &network.get_local_peer().private_key,
+        ));
+        debug!(
+            "{:?}: my public key hash is {}",
+            network.get_local_peer(),
+            &my_pubkey_hash
+        );
```
Sure

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+                    continue;
+                }
+                Err(e) => {
+                    debug!(
```
Sure

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+
+        // update our frontier knowledge
+        for (nkey, new_neighbor) in self.new_frontier.drain() {
+            debug!(
+                "{:?}: Add to frontier of {:?}: {:?}",
+                network.get_local_peer(),
+                &self.cur_neighbor.addr,
+                &nkey
+            );
```
Sure

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+        self.cur_neighbor.in_degree = 1;
+        self.cur_neighbor.out_degree = self.frontier.len() as u32;
+
+        for (_, neighbor_list) in self.resolved_getneighbors_neighbors.iter() {
+            for na in neighbor_list {
+                if na.addrbytes == self.cur_neighbor.addr.addrbytes
+                    && na.port == self.cur_neighbor.addr.port
+                {
+                    self.cur_neighbor.in_degree += 1;
+                }
+            }
+        }
```
Sure thing

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+        }
+
+        // only save if the neighbor is routable from us
+        if self.walk_outbound {
```
Actually, I'm inclined to remove this check.  We should save all discovered neighbors, because we can still walk to them even if they're not routable (i.e. we'd do so via an inbound neighbor walk).

jcnelson replied:
On second thought, we don't want to save a neighbor if (1) it's the first neighbor the walk talks to and (2) the neighbor is inbound.  This is because the `frontier` table contents are served to other neighbors in response to a `GetNeighbors` request.  It would do us no good if the `frontier` table got populated with unroutable entries.  To avoid this, we'd avoid saving `cur_neighbor` if it might be unroutable (which only occurs if (1) and (2) are both true).  However, after stepping from `cur_neighbor` to a subsequent neighbor, the subsequent neighbor is definitely routable from us because we would have already handshaked with it (so we can and should save it).

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2342 @@
+        frontier: &HashMap<NeighborKey, Neighbor>,
+        exclude: Option<&Neighbor>,
+    ) -> Option<Neighbor> {
+        let mut rnd = thread_rng();
+
+        let sample = rnd.gen_range(0, frontier.len());
+        let mut count = 0;
+
+        for (nk, n) in frontier.iter() {
+            count += match exclude {
+                None => 1,
+                Some(ref e) => {
+                    if (*e).addr == *nk {
+                        0
+                    } else {
+                        1
+                    }
+                }
+            };
+            if count >= sample {
+                return Some(n.clone());
+            }
+        }
+        return None;
```
Yup, looks like it.  Thanks for the fix!

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2003 @@
+                Self::pick_random_neighbor(&self.frontier, None).expect("BUG: empty frontier size"); // won't panic since self.frontier.len() > 0
+            let walk_prob: f64 = rnd.gen();
+            if walk_prob
+                < fmin!(
+                    1.0,
+                    self.degree_ratio(network, &self.cur_neighbor, &next_neighbor)
+                )
```
Nifty

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2003 @@
+                }
+            } else {
+                // lost the coin toss. will not take a step
+                Some(self.cur_neighbor.clone())
```
Not in this case.  We're using a Metropolis-Hastings random walk with delayed acceptance.  The "delayed acceptance" part means that we first flip a coin to decide whether to step at all (in this case, the probability of winning is proportional to the ratio of the current node's degree to the next node's degree -- we're more likely to step from a node of high degree to a node of low-degree).  If we don't step, then try again with a different neighbor.

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2003 @@
+            }
+        }
+
+        self.network_pingbacks.clear();
+        self.network_pingbacks.extend(still_pending);
```
Doesn't this create unnecessarily heap allocations?  `.clear()` preserves the allocated memory, and `.extend()` just moves the data.

jcnelson replied:
I'll just use your recommendation.

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2003 @@
+        for (naddr, message) in self.comms.collect_replies(network).into_iter() {
+            // if we got back a HandshakeAccept, and it's on the same chain as us, we're good!
+            match message.payload {
+                StacksMessageType::HandshakeAccept(ref data) => {
+                    debug!("{:?}: received HandshakeAccept from peer {:?}; now known to be routable from us", network.get_local_peer(), &message.to_neighbor_key(&data.handshake.addrbytes, data.handshake.port));
+
+                    let peer_nk =
+                        message.to_neighbor_key(&data.handshake.addrbytes, data.handshake.port);
+                    if !Self::check_handshake_pubkey_hash(&peer_nk, data, &naddr) {
+                        continue;
+                    }
+
+                    self.neighbor_db.add_or_schedule_replace_neighbor(
+                        network,
+                        &naddr,
+                        &message.preamble,
+                        &data.handshake,
+                        None,
+                        &mut self.neighbor_replacements,
+                    )?;
+                }
+                StacksMessageType::StackerDBHandshakeAccept(ref data, ref db_data) => {
+                    debug!("{:?}: received StackerDBHandshakeAccept from peer {:?}; now known to be routable from us", network.get_local_peer(), &message.to_neighbor_key(&data.handshake.addrbytes, data.handshake.port));
+
+                    let peer_nk =
+                        message.to_neighbor_key(&data.handshake.addrbytes, data.handshake.port);
+                    if !Self::check_handshake_pubkey_hash(&peer_nk, data, &naddr) {
+                        continue;
+                    }
+
+                    self.neighbor_db.add_or_schedule_replace_neighbor(
+                        network,
+                        &naddr,
+                        &message.preamble,
+                        &data.handshake,
+                        Some(db_data),
+                        &mut self.neighbor_replacements,
+                    )?;
+                }
```
Sure

jcnelson commented on src/chainstate/coordinator/mod.rs:
```diff
@@ -503,8 +503,12 @@ impl<'a, T: BlockEventDispatcher, U: RewardSetProvider, B: BurnchainHeaderReader
         let atlas_config = atlas_config.unwrap_or(AtlasConfig::new(false));
-        let atlas_db =
-            AtlasDB::connect(atlas_config.clone(), &format!("{}/atlas", path), true).unwrap();
+        let atlas_db = AtlasDB::connect(
+            atlas_config.clone(),
+            &format!("{}/atlas.sqlite", path),
+            true,
+        )
+        .unwrap();
```
It shouldn't.  This path does not correspond to the path created in `TestPeer`, so each unit test created two Atlas DBs -- an empty one used by the coordinator, and the real one used by the peer network test framework.  This forces the coordinator to open the one used by the peer network test framework, which is what I think we wanted all along (i.e. so the coordinator can update it with new attachment data).

jcnelson commented on src/net/neighbors/walk.rs:
```diff
@@ -0,0 +1,2003 @@
+        // advance state!
+        self.set_state(network.get_local_peer(), NeighborWalkState::Finished)?;
+
+        // calculate the walk result
+        if let Some(next_neighbor) = self.next_neighbor.take() {
+            // stepped! return the result
+            return Ok(Some(self.reset(
+                network.get_local_peer(),
+                next_neighbor,
+                false,
+            )));
+        }
```
I think this is a bug.  What ought to happen is `self.walk_outbound` should remain stable over the course of the walk's lifetime, since this value indicates whether or not the first-ever neighbor stepped to was an inbound neighbor (which may or may not be NAT'ed) or an outbound neighbor (which is not NAT'ed from our vantage point).  We should be passing `self.next_walk_outbound` instead of `false`, and `self.next_walk_outbound` should be set in `step()` if we couldn't take any action in MHRWDA.

jcnelson commented on src/net/neighbors.rs:
```diff
@@ -1562,6 +1889,43 @@ impl NeighborWalk {
+    ) -> Result<(), net_error> {
+        let mut tx = network.peerdb.tx_begin()?;
+        let mut neighbor_from_handshake = Neighbor::from_handshake(
+            &mut tx,
```
It does not. Will fix.

jcnelson commented:
Thanks for all the feedback @kantai!  I've addressed all the points.

kantai reviewed (approved):
Thanks, LGTM!

obycode commented on stackslib/src/net/neighbors/comms.rs:
```diff
@@ -0,0 +1,673 @@
+impl ToNeighborKey for NeighborAddress {
+    fn to_neighbor_key(&self, network: &PeerNetwork) -> NeighborKey {
+        // NOTE: PartialEq and Hash for NeighborKey ignore the low bits of peer version
+        // and ignore network ID, and the CovnersationP2P ensures that we never even connect
```
```suggestion
        // and ignore network ID, and the ConversationP2P ensures that we never even connect
```

jcnelson replied:
Thanks!

obycode commented on stackslib/src/net/neighbors/db.rs:
```diff
@@ -0,0 +1,547 @@
+}
+
+impl NeighborWalkDB for PeerDBNeighborWalk {
+    /// implements get_fresh_random_neighbors
```
Are these "implements ..." comments really helpful? Are they just meant to be placeholders?

jcnelson replied:
Will remove -- they're an artifact from an earlier version of this code

obycode reviewed (approved):
LGTM. I really like this refactoring! Just had a few very minor comments.

jcnelson commented on stackslib/src/net/mod.rs:
```diff
@@ -3088,8 +3126,9 @@ pub mod test {
+            let indexer = self.indexer.take().unwrap();
 
-            let indexer = BitcoinIndexer::new_unit_test(&self.config.burnchain.working_dir);
+            // let indexer = BitcoinIndexer::new_unit_test(&self.config.burnchain.working_dir);
```
Yup!

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
